### PR TITLE
HashKeyedCache

### PR DESCRIPTION
**Pull request by Tilps**
Has less pointer chasing than the previous version.

This comes from PR1483.

On my machine I've gotten mixed results.
Simple 34 position 10 second a move benchmark (with appropriate command line args) was almost 2% faster.  But go nodes start position with two different cache sizes were both slightly slower. 
I think it should do better on 'tcec like machines' but I haven't benched there yet.

### CODE REVIEW

**Comment by Tilps** [comment]
I'll do some more testing, but the bugs would have clearly cause performance issues - so I'm marking this ready for review on the basis that I assume the performance is no longer 'mixed'.

**Comment by Tilps** [comment]
My additional nps testing was still good gains, possibly better than before on benchmark. The startposition long think was in the noise.
On some profiles the cpu usage shows up lower with this patch in place. So I think I'm happy with it overall.

**Comment by Tilps** [comment]
Synced in the other changes and retested yet again.
This time I also added a test case for rebuilding search tree for same search after truncating it in between.  Was 7% faster with the newcache implementation.  Start position tests other than that one continue to be 'in the noise, slightly lower'.  Benchmark was over 2% faster this time (closer to 2.5%).
Not a super cohesive story as to when its going to be faster (cache inserts/evicts, cache hits, or cache misses), but overall still looks good.

File: src/utils/cache.h [hunk 1]
```diff
@@ -186,121 +207,99 @@ class LruCache {
...
         : key(key), value(std::move(value)) {}
-    K key;
+    uint64_t key;
     std::unique_ptr<V> value;
     int pins = 0;
-    Item* next_in_hash = nullptr;
-    Item* prev_in_queue = nullptr;
-    Item* next_in_queue = nullptr;
+    // 1 for populated.
+    int metadata = 0;
```
Comment (mooskagh) [inline src/utils/cache.h:222]: Why not `bool` or `enum`?
And probably something like `is_used` is better than `metadata`.

File: src/utils/cache.h [hunk 2]
```diff
@@ -186,121 +207,99 @@ class LruCache {
...
-    iter->next_in_queue = lru_head_;
-    iter->prev_in_queue = nullptr;
-
-    if (lru_head_) {
-      lru_head_->prev_in_queue = iter;
-    }
-    lru_head_ = iter;
-    if (lru_tail_ == nullptr) {
-      lru_tail_ = iter;
+  void ShrinkToCapacity(int capacity) REQUIRES(mutex_) {
```
Comment (mooskagh) [inline src/utils/cache.h:271]: Maybe `EvictToCapacity`?
(I know it was in the original code, but now as I read it, Evict sounds clearer)

File: src/utils/cache.h [hunk 3]
```diff
@@ -31,116 +31,132 @@
...
-    // Now lookup in active list.
-    auto hash = hasher_(key) % hash_.size();
-    for (Item* iter = hash_[hash]; iter; iter = iter->next_in_hash) {
-      if (key == iter->key && value == iter->value.get()) {
-        assert(iter->pins > 0);
-        --iter->pins;
+    // Now the main list.
+    size_t idx = key % hash_.size();
+    while (true) {
+      if (hash_[idx].metadata == 0) break;
```
Comment (mooskagh) [inline src/utils/cache.h:152]: `assert(hash_[idx].metadata == 1)` ?

File: src/utils/cache.h (hunk 1 above)
Comment (Tilps) [reply src/utils/cache.h:222]: Yeah this evolved - it was a tri state and I was just being lazy.  Now a bool seems fine - will change.

File: src/utils/cache.h (hunk 3 above)
Comment (Tilps) [reply src/utils/cache.h:152]: Once its a bool - that'll be implicitly covered by the assert(false) outside the loop.

File: src/utils/cache.h (hunk 1 above)
Comment (Tilps) [reply src/utils/cache.h]: Done.

File: src/utils/cache.h (hunk 2 above)
Comment (Tilps) [reply src/utils/cache.h]: Done.

**Review by mooskagh — APPROVED** [review]

**Pull request merged**